ClientRuntimes/CSharp/Microsoft.Rest.ClientRuntime/BatchDelegatingHandler.cs

```diff
@@ -0,0 +1,185 @@
+            {
+                throw new Exception("Batch response returned unexpected content type.");
+            }
+
+            // pull out the responses
+            MultipartMemoryStreamProvider responseContents = await batchResponse.Content.ReadAsMultipartAsync();
+            if (responseContents == null)
+            {
+                throw new Exception("Failed to convert response content into multipart components.");
+            }
+
+            if (responseContents.Contents.Count != this.requests.Count)
+            {
+                throw new Exception("Batch response returned " + responseContents.Contents.Count + " number of responses instead of the expected " + this.requests.Count + ".");
+            }
+
+            // copy the responses over into the array that is accessible by the tasks waiting for individual responses
+            this.responses = new HttpContent[responseContents.Contents.Count];
+            responseContents.Contents.CopyTo(this.responses, 0);
+
+            // signal tasks that they can proceed with using the response
+            this.batchCompleted.Set();
+        }
+
+        /// <summary>
+        /// Queue an HTTP request that will be eventually sent to the inner handler when you call IssueBatch.
+        /// DO NOT await nor access the result of the returned task until IssueBatch has been successfully awaited.
+        /// </summary>
+        /// <param name="request">HTTP request message to send to the server</param>
+        /// <param name="cancellationToken">Not used</param>
+        /// <returns>Task that when awaited will provide the HTTP response from the server</returns>
+        protected override Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+        {
+            // Sanity check. There must not be any queued responses from the server.
+            if (this.responses != null)
+            {
+                throw new Exception("This batch handler instance has been reused without calling reset first.");
+            }
+
+            // what is my queue position?
+            int position = this.requests.Count;
+
+            // add the request to the queue
+            this.requests.Add(request);
+
+            // send back a task that will return the appropriate value
+            Task<HttpResponseMessage> task = Task<HttpResponseMessage>.Factory.StartNew(() => this.GetResponse(position));
+            return task;
+        }
+
+        /// <summary>
+        /// Wait for the batch call to finish and then return the requested response.
+        /// </summary>
+        /// <param name="index">Which response to return</param>
+        /// <returns>The HTTP response message inside the batch at the given index</returns>
+        private HttpResponseMessage GetResponse(int index)
+        {
+            this.batchCompleted.WaitOne();
+            return this.responses[index].ReadAsHttpResponseMessageAsync().Result;
```

Comment:
What happens if the user:

1. Creates some requests (request set A)
2. Calls `IssueBatch()`
3. Calls `Reset()` before awaiting the responses
4. Creates more requests (request set B)
5. Tries to retrieve the response for a request in set A

From what I can tell, this will:
- Retrieve a response for the request in set B that matches the index of the response from set A
- Throw an IndexOutOfRange exception if you try to retrieve the response from a request from set B that has a higher index that the `count()` of set A responses

I'm not sure that this is the behavior, but from looking at the code, this is what I believe will happen.

I don't know of an easy solution for this issue off the top of my head. My instincts would be to create a private class like `BatchContext` that stores the responses and requests each time `IssueBatch()` is called. That way the `requests` collection can be guaranteed to be the same when the responses are read, while allowing more requests to be created.